File: SaidIt/src/androidTest/java/eu/mrogalski/saidit/SaidItFragmentTest.java
package eu.mrogalski.saidit;

import android.Manifest;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.SystemClock;
import android.view.View;
import androidx.test.core.app.ApplicationProvider;
import androidx.test.core.app.ActivityScenario;
import androidx.test.espresso.UiController;
import androidx.test.espresso.ViewAction;
import androidx.test.espresso.matcher.RootMatchers;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.rule.GrantPermissionRule;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import java.util.concurrent.atomic.AtomicReference;
import org.hamcrest.Matcher;
import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.closeSoftKeyboard;
import static androidx.test.espresso.action.ViewActions.replaceText;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.isRoot;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.junit.Assert.assertNotNull;

import com.siya.epistemophile.R;

@RunWith(AndroidJUnit4.class)
public class SaidItFragmentTest {

    private static final String PREF_FIRST_RUN = "is_first_run";
    private static final String PREF_TOUR_NEXT = "show_tour_on_next_launch";

    @Rule
    public final GrantPermissionRule permissionRule = GrantPermissionRule.grant(
            Manifest.permission.RECORD_AUDIO,
            Manifest.permission.POST_NOTIFICATIONS,
            Manifest.permission.FOREGROUND_SERVICE,
            Manifest.permission.FOREGROUND_SERVICE_MICROPHONE
    );

    @Before
    public void configurePreferences() {
        Context context = ApplicationProvider.getApplicationContext();
        SharedPreferences preferences = context.getSharedPreferences(SaidIt.PACKAGE_NAME, Context.MODE_PRIVATE);
        preferences.edit()
                .putBoolean(PREF_FIRST_RUN, false)
                .putBoolean(PREF_TOUR_NEXT, false)
                .apply();
        SaidItActivity.setTestSkipPermissionRequest(true);
    }

    @After
    public void resetFlags() {
        SaidItActivity.setTestSkipPermissionRequest(false);
    }

    @Test
    public void testSaveClipFlow_showsProgressDialog() {
        ActivityScenario<SaidItActivity> scenario = ActivityScenario.launch(SaidItActivity.class);
        SaidItService service = awaitBoundService(scenario);
        service.setTestEnvironment(true);
        service.setTestAutoSaveDelayMs(400L);
        service.enableListening();

        onView(withId(R.id.save_clip_button)).perform(click());
        onView(withId(R.id.recording_name)).perform(replaceText("Test clip"), closeSoftKeyboard());
        onView(withId(R.id.save_button)).perform(click());
        onView(isRoot()).perform(waitFor(250));

        onView(withText("Saving Recording"))
                .inRoot(RootMatchers.isDialog())
                .check(matches(isDisplayed()));

        scenario.close();
    }

    private SaidItService awaitBoundService(ActivityScenario<SaidItActivity> scenario) {
        AtomicReference<SaidItService> reference = new AtomicReference<>();
        long start = SystemClock.uptimeMillis();
        long timeout = 15_000L;

        while (reference.get() == null && SystemClock.uptimeMillis() - start < timeout) {
            scenario.onActivity(activity -> {
                activity.ensureServiceBoundForTest();
                reference.set(activity.getEchoService());
            });
            if (reference.get() != null) {
                break;
            }
            SystemClock.sleep(50);
        }

        SaidItService service = reference.get();
        assertNotNull("Service should be bound before interacting with UI", service);
        return service;
    }

    private ViewAction waitFor(long millis) {
        return new ViewAction() {
            @Override
            public Matcher<View> getConstraints() {
                return isRoot();
            }

            @Override
            public String getDescription() {
                return "wait for " + millis + " milliseconds";
            }

            @Override
            public void perform(UiController uiController, View view) {
                uiController.loopMainThreadForAtLeast(millis);
            }
        };
    }
}

File: SaidIt/src/androidTest/kotlin/eu/mrogalski/saidit/SaidItFragmentTest.kt
package eu.mrogalski.saidit

import android.Manifest
import android.content.Context
import android.content.SharedPreferences
import androidx.appcompat.app.AlertDialog
import androidx.test.core.app.ActivityScenario
import androidx.test.core.app.ApplicationProvider
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.action.ViewActions.closeSoftKeyboard
import androidx.test.espresso.action.ViewActions.replaceText
import androidx.test.espresso.matcher.ViewMatchers.isRoot
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.rule.GrantPermissionRule
import com.siya.epistemophile.R
import com.siya.epistemophile.testing.AndroidTestUtils.awaitBoundService
import com.siya.epistemophile.testing.AndroidTestUtils.waitFor
import com.siya.epistemophile.testing.AndroidTestUtils.waitUntil
import org.junit.*
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class SaidItFragmentTest {

    companion object {
        private const val PREF_FIRST_RUN = "is_first_run"
        private const val PREF_TOUR_NEXT = "show_tour_on_next_launch"
    }

    @get:Rule
    val permissionRule: GrantPermissionRule = GrantPermissionRule.grant(
        Manifest.permission.RECORD_AUDIO,
        Manifest.permission.POST_NOTIFICATIONS,
        Manifest.permission.FOREGROUND_SERVICE,
        Manifest.permission.FOREGROUND_SERVICE_MICROPHONE
    )

    @Before
    fun configurePreferences() {
        val context = ApplicationProvider.getApplicationContext<Context>()
        val preferences: SharedPreferences = context.getSharedPreferences(SaidIt.PACKAGE_NAME, Context.MODE_PRIVATE)
        preferences.edit()
            .putBoolean(PREF_FIRST_RUN, false)
            .putBoolean(PREF_TOUR_NEXT, false)
            .apply()
        SaidItActivity.setTestSkipPermissionRequest(true)
    }

    @After
    fun resetFlags() {
        SaidItActivity.setTestSkipPermissionRequest(false)
    }

    @Test
    fun testSaveClipFlow_showsProgressDialog() {
        val scenario = ActivityScenario.launch(SaidItActivity::class.java)
        val service = awaitBoundService(scenario)
        service.isTestEnvironment = true
        service.testAutoSaveDelayMs = 400L
        service.enableListening()

        onView(withId(R.id.save_clip_button)).perform(click())
        onView(withId(R.id.recording_name)).perform(replaceText("Test clip"), closeSoftKeyboard())
        onView(withId(R.id.save_button)).perform(click())

        val appeared = waitUntil(scenario, timeoutMs = 5_000) { activity ->
            val fragment = activity.supportFragmentManager.findFragmentById(R.id.container) as? SaidItFragment
            val dialog: AlertDialog? = fragment?.let { findProgressDialog(it) }
            dialog?.isShowing == true
        }
        assertTrue("Progress dialog should appear and be showing", appeared)

        scenario.close()
    }
}

private fun findProgressDialog(fragment: SaidItFragment): AlertDialog? {
    val candidates = arrayOf(
        "getProgressDialogForTest",
        "getProgressDialogForTest\$SaidIt_debug",
        "getProgressDialogForTest\$SaidIt_release"
    )
    for (name in candidates) {
        try {
            val method = fragment::class.java.getMethod(name)
            val result = method.invoke(fragment)
            if (result is AlertDialog) return result
        } catch (_: Exception) {
            // try next
        }
    }
    return null
}
